src/Chainweb/Pact/PactService.hs

```diff
@@ -464,45 +479,37 @@ minerReward bh = do
 
 -- | Note: The BlockHeader param here is the PARENT HEADER of the new
 -- block-to-be
+--
+-- Precondition: restoreCheckpointer
+-- Postcondition: finalizeCheckpointer discard
+--
 execNewBlock
     :: PayloadCas cas
     => MemPoolAccess
     -> BlockHeader
     -> Miner
     -> PactServiceM cas PayloadWithOutputs
 execNewBlock mpAccess parentHeader miner = withDiscardedBatch $ do
-    let pHeight = _blockHeight parentHeader
-        pHash = _blockHash parentHeader
-        bHeight = succ pHeight
-
-    let rewindPoint = Just (bHeight, pHash)
-
-    -- TODO: should we work towards uncommenting this? Currently, cutdb traffic
-    -- moves the pactdb "current block" cursor too often for us to be sure that
-    -- the miner and consensus are going to be in sync for newBlock. Changes to
-    -- consensus or to checkpoint API may allow us to re-enable this check
-    --
-    -- cp <- view (psCheckpointEnv . cpeCheckpointer)
-    -- latest <- liftIO $ getLatestBlock cp
-    -- when (latest /= Just (pHeight, pHash)) $
-    --     throwM $ PactServiceIllegalRewind rewindPoint latest
-
-    -- rewind should usually be trivial / no-op
-    rewindTo rewindPoint $ \pdbenv -> do
-        logInfo $ "execNewBlock, about to get call processFork: "
-               <> " (parent height = " <> sshow pHeight <> ")"
-               <> " (parent hash = " <> sshow pHash <> ")"
-        liftIO $ mpaProcessFork mpAccess parentHeader
-        liftIO $ mpaSetLastHeader mpAccess parentHeader
-        cp <- view (psCheckpointEnv . cpeCheckpointer)
-        let validate = validateChainwebTxsPreBlock pdbenv cp
-        newTrans <- liftIO $
-                    mpaGetBlock mpAccess validate bHeight pHash parentHeader
-        -- locally run 'execTransactions' with updated blockheight data
-        results <- withParentBlockData parentHeader $
-          execTransactions (Just pHash) miner newTrans pdbenv
-
-        return $! toPayloadWithOutputs miner results
+    pdbenv <- rewindTo rewindPoint
+    logInfo $ "execNewBlock, about to get call processFork: "
+            <> " (parent height = " <> sshow pHeight <> ")"
+            <> " (parent hash = " <> sshow pHash <> ")"
+    liftIO $ mpaProcessFork mpAccess parentHeader
+    liftIO $ mpaSetLastHeader mpAccess parentHeader
+    cp <- view (psCheckpointEnv . cpeCheckpointer)
+    let validate = validateChainwebTxsPreBlock pdbenv cp
+    newTrans <- liftIO $
+        mpaGetBlock mpAccess validate bHeight pHash parentHeader
+    -- locally run 'execTransactions' with updated blockheight data
+    results <- withParentBlockData parentHeader $
+        execTransactions (Just pHash) miner newTrans pdbenv
+    finalizeCheckpointer discard
```

Comment:
This call is new. Without it the function would leave an `Block` savepoint on the sqlite transaction stack.